File: web/navigation.js
import {escape as e,position as positionBadge} from './ui.js';

const triangle=up=>`<svg viewBox="0 0 10 10" aria-hidden="true"><path d="${up?'M5 2 9.5 8h-9z':'M5 8 .5 2h9z'}"/></svg>`;
const burger='<svg viewBox="0 0 10 10" aria-hidden="true"><path d="M1 2h8M1 5h8M1 8h8" fill="none" stroke="currentColor" stroke-width="1.4" stroke-linecap="round"/></svg>';

// A small block for the left of a page header: a triangle up to the previous member of the group, a menu listing every member, a triangle down
// to the next one. `href` builds the link of a member, `label` names it in text and `row` in the list. The ends of the group have no link on
// their side; the greyed triangle keeps the block steady. A group of one has nothing to step through.
function neighbours(nav,{href,scope,listLabel,label=item=>item.name,row=item=>e(label(item))}){
 if(!nav||nav.total<2)return '';
 const step=(item,direction,name)=>item
  ?`<a class="entity-step ${direction}" href="${e(href(item))}" rel="${direction}" aria-label="${name} : ${e(label(item))}" title="${name} : ${e(label(item))}">${triangle(direction==='prev')}</a>`
  :`<span class="entity-step ${direction}" aria-hidden="true">${triangle(direction==='prev')}</span>`;
 const caption=`${scope} · ${nav.index+1} / ${nav.total}`;
 const rows=nav.items.map((item,index)=>`<li><a href="${e(href(item))}"${index===nav.index?' aria-current="true"':''}>${row(item)}</a></li>`).join('');
 return `<div class="entity-nav" role="group" aria-label="${e(listLabel)}">${step(nav.previous,'prev','Précédent')}<details class="entity-menu"><summary aria-label="${e(listLabel)}" title="${e(caption)}">${burger}</summary><div class="entity-menu-panel"><p class="entity-menu-scope">${e(caption)}</p><ul>${rows}</ul></div></details>${step(nav.next,'next','Suivant')}</div>`;
}

// Clubs of a division, alphabetically; a club outside any division steps through the clubs of its country. Moving keeps the open tab.
// Homonymous clubs (the source lists a club and its empty duplicate) carry their squad size, so that a step never looks like a no-op.
export function clubNavigation(nav,section){
 if(!nav)return '';
 const division=nav.scope.kind==='division';
 const seen=new Set(),twins=new Set();for(const club of nav.items)(seen.has(club.name)?twins:seen).add(club.name);
 return neighbours(nav,{href:club=>`#/club/${club.id}${section?`/${section}`:''}`,scope:division?nav.scope.name:`Tous les clubs · ${nav.scope.name}`,
  listLabel:`Choisir un club${division?` de ${nav.scope.name}`:` · ${nav.scope.name}`}`,
  label:club=>twins.has(club.name)?`${club.name} (${club.squad} joueur${club.squad>1?'s':''})`:club.name});
}

// The nations of the same confederation. Moving keeps the open tab.
export function nationNavigation(nav,tab) {
 return nav?neighbours(nav,{href:nation=>`#/international/nation/${nation.id}${tab?`/${tab}`:''}`,scope:`Confédération · ${nav.scope.name}`,
  listLabel:`Choisir une nation · ${nav.scope.name}`}):'';
}

// The squad of the player's club, goalkeepers first, each with the colour of its position.
export function playerNavigation(nav){
 return nav?neighbours(nav,{href:player=>`#/player/${player.id}`,scope:`Effectif · ${nav.scope.name}`,listLabel:`Choisir un joueur de ${nav.scope.name}`,
  row:player=>`${positionBadge(player.position)}<span>${e(player.name)}</span>`}):'';
}

// The divisions of a country from the top down, then its cup. A competition page has no tab in common with the next one, so it opens on its default.
export function competitionNavigation(nav){
 return nav?neighbours(nav,{href:competition=>`#/league/${competition.id}`,scope:`Compétitions · ${nav.scope.name}`,listLabel:`Choisir une compétition · ${nav.scope.name}`}):'';
}

File: src/api/navigation.py
"""Neighbours and pick lists for stepping between the members of a group; read-only projection."""
from core.domain.world import World
from . import views as v

COMPETITION_KINDS = {"league": 0, "cup": 1, "europe": 2}


def navigation(items: list[dict], current_id: int, scope: dict) -> dict:
    """`items` is the whole group in display order; the ends have no previous or next."""
    index = next(position for position, item in enumerate(items) if item["id"] == current_id)
    return {"scope": scope, "index": index, "total": len(items), "items": items,
            "previous": items[index - 1] if index > 0 else None,
            "next": items[index + 1] if index + 1 < len(items) else None}


def club_navigation(world: World, club_id: int) -> dict:
    """The clubs of the same division, or of the same country when the club plays in none."""
    club = world.clubs[club_id]
    if club.competition_id is not None:
        members = [item for item in world.clubs.values() if item.competition_id == club.competition_id]
        scope = {"kind": "division", "id": club.competition_id, "name": world.competitions[club.competition_id].name}
    else:
        members = [item for item in world.clubs.values() if item.nation == club.nation]
        scope = {"kind": "country", "code": club.nation, "name": world.nation_names.get(club.nation, club.nation)}
    members.sort(key=lambda item: (v.normalized(item.name), item.id))
    # The squad size tells apart the homonyms of the source data (a club and its empty duplicate).
    return navigation([{"id": item.id, "name": item.name, "squad": len(item.player_ids)} for item in members], club_id, scope)


def player_navigation(world: World, player_id: int) -> dict | None:
    """The squad of the player's club, goalkeepers first as in the squad table; none for retirees and free agents."""
    if player_id in world.retired:
        return None
    club = world.clubs.get(world.players[player_id].club_id)
    if club is None:
        return None
    squad = sorted((world.players[pid] for pid in club.player_ids),
                   key=lambda item: (v.position_rank(item.position.value), v.normalized(item.name), item.id))
    return navigation([{"id": item.id, "name": item.name, "position": item.position.value} for item in squad],
                      player_id, {"kind": "club", "id": club.id, "name": club.name})


def nation_navigation(world: World, nation_id: int) -> dict:
    """The nations of the same confederation, alphabetically."""
    team = world.international.nations[nation_id]
    members = [item for item in world.international.nations.values() if item.federation == team.federation]
    members.sort(key=lambda item: (v.normalized(item.name), item.id))
    return navigation([{"id": item.id, "name": item.name} for item in members], nation_id,
                      {"kind": "federation", "name": team.federation})


def competition_navigation(world: World, competition_id: int) -> dict:
    """The competitions of the same country: divisions from the top down, then the cup."""
    competition = world.competitions[competition_id]
    members = sorted((item for item in world.competitions.values() if item.nation == competition.nation),
                     key=lambda item: (COMPETITION_KINDS.get(item.kind, len(COMPETITION_KINDS)), item.level, item.code or "", item.id))
    return navigation([{"id": item.id, "name": item.name, "kind": item.kind} for item in members], competition_id,
                      {"kind": "country", "code": competition.nation, "name": world.nation_names.get(competition.nation, competition.nation)})
